### ISSUE TITLE
python3 doesn't like pickling I3Frames (Trac #1923)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1915]
Typical error is:

```text
Traceback (most recent call last):
  File "/build/buildslave/paimon/Arch_Linux/source/recclasses/resources/test/test_suite.py", line 28, in test_serialization
    pickle.dump(frame, pseudo_file)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x97 in position 41: invalid start byte
```

with the filename, byte, and position obviously being different. 

Affects the following  tests

```text
full_event_followup::test_json_packing_and_unpacking.py
icetray::test_python_frame_dumps_loads.py
recclasses::test_suite.py
```

Migrated from https://code.icecube.wisc.edu/ticket/1923
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:12:54", 
    "description": "Typical error is:\n\n{{{\nTraceback (most recent call last):\n  File \"/build/buildslave/paimon/Arch_Linux/source/recclasses/resources/test/test_suite.py\", line 28, in test_serialization\n    pickle.dump(frame, pseudo_file)\nUnicodeDecodeError: 'utf-8' codec can't decode byte 0x97 in position 41: invalid start byte\n}}}\n\nwith the filename, byte, and position obviously being different. \n\nAffects the following  tests\n{{{\nfull_event_followup::test_json_packing_and_unpacking.py\nicetray::test_python_frame_dumps_loads.py\nrecclasses::test_suite.py\n}}}", 
    "reporter": "nega", 
    "cc": "olivas", 
    "resolution": "fixed", 
    "_ts": "1550067174476394", 
    "component": "combo reconstruction", 
    "summary": "python3 doesn't like pickling I3Frames", 
    "priority": "normal", 
    "keywords": "icetray, recclasses, full_event_followup, testing, python", 
    "time": "2016-12-15T20:40:27", 
    "milestone": "", 
    "owner": "jpa14", 
    "type": "enhancement"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20161216T11:31:10: kjmeagher commented:

> The problem is that I3Frame is outputting an array of `unsigned char` with values greater than `0x80` which python is then interpreting as a string. In python2 this works fine, but in python3 it tries to encode this as utf-8 to get bytes and fails. I think the answer is an `#ifdef python3` which used `PyMemoryView_FromMemory` to export the array.

**Comment by icecube-trac** [comment]
Trac update at 20161221T00:48:46: 
* jpa14 changed status from "new" to "assigned"
* jpa14 changed owner from "" to "jpa14"

**Comment by icecube-trac** [comment]
Trac update at 20161221T00:50:31: 
* olivas changed milestone from "December 2016" to "January 2017"
* olivas changed type from "defect" to "enhancement"
* olivas commented:

> I'd call this an enhancement, since we haven't officially adopted Python3...soon though...hopefully.  Another thing to talk about next year.

**Comment by icecube-trac** [comment]
Trac update at 20161221T17:48:52: flauber commented:

> Sorry to ask, but just looking at the test_suit.py, you are using StringIO to pickle to.
> I always thought pickle writes bytes, StringIO does not support bytes but strings.
> On Python2, the difference is not notable but after the python3 started to separate bytes and strings, I guess one should be using BytesIO instead?

**Comment by icecube-trac** [comment]
Trac update at 20170111T20:46:13: david.schultz commented:

> Re StringIO, that's correct and will be part of this patch.
> 
> As for the MemoryView:
> `TypeError: can't pickle memoryview objects`
> So that's not the correct way to do it, or at least not the whole solution.

**Comment by icecube-trac** [comment]
Trac update at 20170111T22:27:49: 
* david.schultz changed resolution from "" to "fixed"
* david.schultz commented:

> In changeset 2768:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2768"
> pickling support in python3. fixes #1915
> ```


* david.schultz changed status from "assigned" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20190111T20:52:39: nwhitehorn commented:

> In changeset 168784:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="168784"
> pickling support in python3. fixes #1915
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:12:54: 
* olivas changed milestone from "January 2017" to ""
* olivas commented:

> Milestone January 2017 deleted
